### PR TITLE
feat: Use ERC-2258 in PoolFDT to allow for farming + interest (SC-2046)

### PR DESCRIPTION
**Pull request by lucas-manuel**
# Description

This PR will allow for PoolFDTs to be used in MplRewards without the LPs ever giving up custody. Follows the [ERC-2258](https://github.com/ethereum/EIPs/issues/2258) standard. 

WIP, still need to:
- [ ] Finalize functionality, make sure all checks are in place
- [ ] Add same functionality to StakeLockerFDTs
- [ ] Add unit testing for custody allowance functionality for PoolFDT
- [ ] Add unit testing for custody allowance functionality for StakeLockerFDT

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: test.sh [hunk 1]
```diff
@@ -13,4 +13,4 @@ export DAPP_LINK_TEST_LIBRARIES=1
 
 if [ ${1} ]; then match=${1}; dapp_test_verbosity=2; else match="contracts/test"; dapp_test_verbosity=1; fi
 
-LANG=C.UTF-8 dapp test --match "$match" --rpc-url "$ETH_RPC_URL" --verbosity $dapp_test_verbosity --cache "cache/dapp-cache" --fuzz-runs 100
+LANG=C.UTF-8 dapp test --match "$match" --rpc-url "$ETH_RPC_URL" --verbosity $dapp_test_verbosity --cache "cache/dapp-cache" --fuzz-runs 10
```
Comment (satyamakgec) [inline test.sh:16]: do we want to bump up that to 10 again

Comment (lucas-manuel) [reply test.sh:16]: Yeah I want to personally - its starting to take a while locally

File: contracts/Pool.sol [hunk 2]
```diff
@@ -451,6 +455,42 @@ contract Pool is PoolFDT {
...
+        PoolLib.increaseCustodyAllowanceChecks(custodian, amount, newTotalAllowance, balanceOf(msg.sender));
+        
+        custodyAllowance[msg.sender][custodian] = newAllowance;
+        totalCustodyAllowance[msg.sender]       = newTotalAllowance;
+        emit CustodyAllowanceChanged(msg.sender, custodian, oldAllowance, newAllowance);
+    }
+
+    /**
+        @dev Transfer custodied FDTs to `to` address.
+        @notice sender and receiver should always be equal in this implementation.
```
Comment (michaelelliot) [inline contracts/Pool.sol:477]: Should this be more explicit? e.g. "from and to should always be equal in this implementation."
Sender could be confused with `msg.sender` (the custodian).

File: contracts/library/PoolLib.sol [hunk 3]
```diff
@@ -238,24 +238,42 @@ library PoolLib {
...
+        require(recognizableLosses == uint256(0),                "P:RECOG_LOSSES");                  // If an LP has unrecognized losses, they must recognize losses through withdraw
+        updateDepositDate(depositDate, toBalance, wad, to);                                          // Update deposit date of recipient
+    }
+
+    /**
+        @dev Performs all necessary checks for a `transferByCustodian` call.
+        @dev From and to must always be equal. (TODO: Should we do this?)
+     */
+    function transferByCustodianChecks(address from, address to, uint256 amount, uint256 custodyAllowance) external {
+        require(to == from,                 "P:INVALID_RECEIVER");
```
Comment (michaelelliot) [inline contracts/library/PoolLib.sol:265]: Should we consider implementing a version of `transferByCustodian` where `to != from`?

We're not ever going to be transferring tokens away in `MplRewards` but there may be an integration down the road where this is the case.

**Comment by michaelelliot** [comment]
Since `MplRewards` never beneficially holds any tokens now, we could remove the `require(tokenAddress != address(stakingToken))` safeguard from `recoverERC20()` in case Pool FDTs are ever accidentally sent directly to the rewards contract.

**Review by deluca-mike — CHANGES_REQUESTED** [review]

File: contracts/Pool.sol (hunk 2 above)
Comment (lucas-manuel) [reply contracts/Pool.sol:467]: Yeah good call

File: contracts/Pool.sol [hunk 4]
```diff
@@ -441,6 +445,42 @@ contract Pool is PoolFDT {
...
+        PoolLib.increaseCustodyAllowanceChecks(custodian, amount, newTotalAllowance, balanceOf(msg.sender));
+        
+        custodyAllowance[msg.sender][custodian] = newAllowance;
+        totalCustodyAllowance[msg.sender]       = newTotalAllowance;
+        emit CustodyAllowanceChanged(msg.sender, custodian, oldAllowance, newAllowance);
+    }
+
+    /**
+        @dev Transfer custodied FDTs to `to` address.
+        @notice from and to should always be equal in this implementation.
```
Comment (satyamakgec) [inline contracts/Pool.sol:467]: Sorry for the last pointer but we can update the dev comments here to clear it more for the external auditors.

**Review by satyamakgec — APPROVED** [review]

**Review by satyamakgec — DISMISSED** [review]

**Review by deluca-mike — APPROVED** [review]

**Pull request merged**